File: android/app/src/test/java/com/nexusriot/rezoagwe/KvStoreTest.kt
package com.nexusriot.rezoagwe

import com.nexusriot.rezoagwe.core.KvStore
import com.nexusriot.rezoagwe.core.WriteOptions
import com.nexusriot.rezoagwe.proto.Digest
import com.nexusriot.rezoagwe.proto.KVAction
import com.nexusriot.rezoagwe.proto.KVUpdate
import com.nexusriot.rezoagwe.proto.KeyVersion
import com.nexusriot.rezoagwe.proto.Version
import org.junit.Assert.assertEquals
import org.junit.Assert.assertFalse
import org.junit.Assert.assertNotNull
import org.junit.Assert.assertNull
import org.junit.Assert.assertTrue
import org.junit.Test

class KvStoreTest {

    @Test
    fun applyIsLastWriteWins() {
        val kv = KvStore("self")
        kv.write("k", "local")

        assertFalse(
            "an older version must not win",
            kv.apply(KVUpdate(KVAction.SET, "k", "stale", Version(0, "peer"))),
        )
        assertEquals("local", kv["k"])

        assertTrue(kv.apply(KVUpdate(KVAction.SET, "k", "fresh", Version(9, "peer"))))
        assertEquals("fresh", kv["k"])
    }

    @Test
    fun tombstoneBlocksStaleResurrect() {
        val kv = KvStore("self")
        kv.apply(KVUpdate(KVAction.SET, "k", "v", Version(2, "a")))
        kv.apply(KVUpdate(KVAction.DELETE, "k", "", Version(5, "a")))
        assertNull(kv["k"])

        kv.apply(KVUpdate(KVAction.SET, "k", "zombie", Version(3, "b")))
        assertNull("a stale set must not resurrect a deleted key", kv["k"])

        kv.apply(KVUpdate(KVAction.SET, "k", "back", Version(6, "b")))
        assertEquals("back", kv["k"])
    }

    @Test
    fun clockAdvancesOnApply() {
        val kv = KvStore("self")
        kv.apply(KVUpdate(KVAction.SET, "k", "v", Version(100, "peer")))
        val update = kv.write("mine", "x")!!
        assertTrue("local write must sort after everything seen", update.version.counter > 100)
    }

    /** The Go and Kotlin stores must resolve a conflict the same way, or replicas diverge. */
    @Test
    fun concurrentWritesConvergeOnNodeIdTiebreak() {
        val a = KvStore("node-a")
        val b = KvStore("node-b")

        val fromA = a.write("k", "from-a")!!
        val fromB = b.write("k", "from-b")!!
        a.apply(fromB)
        b.apply(fromA)

        assertEquals(a["k"], b["k"])
        assertEquals("from-b", a["k"])
    }

    @Test
    fun compareAndSwap() {
        val kv = KvStore("self")
        val first = kv.write("lock", "alice")!!

        assertNull(
            "CAS against the wrong version must fail",
            kv.write("lock", "bob", WriteOptions(expect = Version(99, "x"))),
        )
        assertEquals("alice", kv["lock"])

        assertNotNull(kv.write("lock", "bob", WriteOptions(expect = first.version)))
        assertEquals("bob", kv["lock"])
    }

    @Test
    fun compareAndSwapAgainstAbsence() {
        val kv = KvStore("self")
        assertNotNull("claiming a free key must succeed", kv.write("lock", "alice", WriteOptions(expect = Version())))
        assertNull("claiming a held key must fail", kv.write("lock", "bob", WriteOptions(expect = Version())))
        assertEquals("alice", kv["lock"])
    }

    @Test
    fun ttlHidesExpiredKey() {
        var now = 1000L
        val kv = KvStore("self") { now }
        kv.write("session", "token", WriteOptions(expiresAt = 1010))

        assertEquals("token", kv["session"])
        now = 1011
        assertNull("an expired key must read as absent", kv["session"])
        assertEquals(0, kv.size())
    }

    /**
     * Expiry must not bump the Lamport clock: a sweep that outranked a concurrent
     * legitimate write would silently destroy it.
     */
    @Test
    fun sweepKeepsVersion() {
        var now = 1000L
        val kv = KvStore("self") { now }
        val written = kv.write("session", "token", WriteOptions(expiresAt = 1001))!!

        now = 1002
        assertEquals(1, kv.sweepExpired())

        val update = kv.updatesFor(listOf("session")).single()
        assertEquals(KVAction.DELETE, update.action)
        assertEquals(written.version, update.version)

        kv.write("session", "fresh")
        assertEquals("a later write must still beat the swept tombstone", "fresh", kv["session"])
    }

    @Test
    fun gcReclaimsOldTombstonesOnly() {
        var now = 1000L
        val kv = KvStore("self") { now }
        kv.write("a", "1")
        kv.remove("a")
        kv.write("b", "2")

        assertEquals(0, kv.gcTombstones(3600))
        now = 1000 + 7200
        assertEquals(1, kv.gcTombstones(3600))
        assertEquals(0, kv.tombstones())
        assertEquals("2", kv["b"])
    }

    @Test
    fun digestPaginatesTheKeyspace() {
        val kv = KvStore("self")
        listOf("a", "b", "c", "d").forEach { kv.write(it, it) }

        val first = kv.digest("", 2)
        assertEquals(listOf("a", "b"), first.entries.map { it.key })
        assertTrue("a truncated digest must report an upper bound", first.hi.isNotEmpty())

        val second = kv.digest("b", 2)
        assertEquals(listOf("c", "d"), second.entries.map { it.key })
        assertEquals("the final batch reaches the end of the keyspace", "", second.hi)
    }

    @Test
    fun reconcilePushesAndPulls() {
        val local = KvStore("local")
        local.apply(KVUpdate(KVAction.SET, "shared", "new", Version(5, "x")))
        local.apply(KVUpdate(KVAction.SET, "only-local", "v", Version(1, "x")))
        local.apply(KVUpdate(KVAction.SET, "stale-local", "old", Version(1, "x")))

        val (push, pull) = local.reconcile(
            Digest(
                from = "peer",
                entries = listOf(
                    KeyVersion("shared", Version(2, "x")),
                    KeyVersion("stale-local", Version(9, "x")),
                    KeyVersion("only-peer", Version(1, "x")),
                ),
            ),
            maxPush = 100,
            maxPull = 100,
        )

        val pushed = push.map { it.key }.toSet()
        assertTrue("shared" in pushed)
        assertTrue("only-local" in pushed)
        assertFalse("must not push what the peer holds newer", "stale-local" in pushed)
        assertEquals(setOf("stale-local", "only-peer"), pull.toSet())
    }

    /** A key outside the advertised range was simply not reached yet, and must not be pushed. */
    @Test
    fun reconcileRespectsDigestRange() {
        val local = KvStore("local")
        listOf("a" to "1", "m" to "2", "z" to "3").forEach { (k, v) -> local.write(k, v) }

        val (push, _) = local.reconcile(
            Digest(from = "peer", lo = "", hi = "n", entries = listOf(KeyVersion("a", Version(99, "x")))),
            maxPush = 100,
            maxPull = 100,
        )
        assertEquals(listOf("m"), push.map { it.key })
    }

    @Test
    fun historyRecordsWriters() {
        val kv = KvStore("self")
        kv.write("k", "v1")
        kv.apply(KVUpdate(KVAction.SET, "k", "v2", Version(50, "peer")))

        val history = kv.history("k")
        assertEquals(2, history.size)
        assertTrue(history[0].local)
        assertFalse(history[1].local)
        assertEquals("v2", history[1].value)
    }

    @Test
    fun digestRangeBoundsMatchGo() {
        // Lo is exclusive (it is the sender's cursor), hi is exclusive.
        assertFalse(KvStore.inDigestRange("a", "a", ""))
        assertTrue(KvStore.inDigestRange("b", "a", ""))
        assertTrue(KvStore.inDigestRange("b", "a", "c"))
        assertFalse(KvStore.inDigestRange("c", "a", "c"))
    }
}

File: pkg/discovery/model/kvstore_test.go
package model

import (
	"testing"
	"time"

	pb "github.com/nexusriot/rezoagwe/pkg/proto"
)

// A remote update wins only if its version is strictly newer than what's held.
func TestApplyLastWriteWins(t *testing.T) {
	kv := NewKVStore(":self")
	kv.Set("k", "local") // local write → version {1, ":self"}

	// Older counter: ignored.
	if kv.Apply(pb.KVUpdate{Action: pb.KVSet, Key: "k", Value: "stale", Version: pb.Version{Counter: 0, Node: ":peer"}}) {
		t.Fatal("Apply of older version reported a change")
	}
	if v, _ := kv.Get("k"); v != "local" {
		t.Fatalf("k = %q, want %q after stale apply", v, "local")
	}

	// Newer counter: wins.
	if !kv.Apply(pb.KVUpdate{Action: pb.KVSet, Key: "k", Value: "fresh", Version: pb.Version{Counter: 9, Node: ":peer"}}) {
		t.Fatal("Apply of newer version reported no change")
	}
	if v, _ := kv.Get("k"); v != "fresh" {
		t.Fatalf("k = %q, want %q after newer apply", v, "fresh")
	}
}

// A tombstone must block a stale set from resurrecting the key, but a newer
// set must still bring it back.
func TestTombstoneBlocksStaleResurrect(t *testing.T) {
	kv := NewKVStore(":self")
	kv.Apply(pb.KVUpdate{Action: pb.KVSet, Key: "k", Value: "v", Version: pb.Version{Counter: 2, Node: ":a"}})
	kv.Apply(pb.KVUpdate{Action: pb.KVDelete, Key: "k", Version: pb.Version{Counter: 5, Node: ":a"}})

	if _, ok := kv.Get("k"); ok {
		t.Fatal("key present after delete")
	}
	kv.Apply(pb.KVUpdate{Action: pb.KVSet, Key: "k", Value: "zombie", Version: pb.Version{Counter: 3, Node: ":b"}})
	if _, ok := kv.Get("k"); ok {
		t.Fatal("stale set resurrected a tombstoned key")
	}
	kv.Apply(pb.KVUpdate{Action: pb.KVSet, Key: "k", Value: "back", Version: pb.Version{Counter: 6, Node: ":b"}})
	if v, ok := kv.Get("k"); !ok || v != "back" {
		t.Fatalf("newer set did not resurrect: (%q, %v)", v, ok)
	}
}

// Applying a remote version must advance the Lamport clock so this node's
// next local write sorts strictly after it.
func TestLamportClockAdvancesOnApply(t *testing.T) {
	kv := NewKVStore(":self")
	kv.Apply(pb.KVUpdate{Action: pb.KVSet, Key: "k", Value: "v", Version: pb.Version{Counter: 100, Node: ":peer"}})

	u := kv.Set("mine", "x")
	if u.Version.Counter <= 100 {
		t.Fatalf("local write counter = %d, want > 100 (clock didn't advance)", u.Version.Counter)
	}
}

// Two nodes writing the same key concurrently must converge to the same value
// on both sides after exchanging updates — the whole point of versioning.
func TestConvergenceOnConcurrentWrite(t *testing.T) {
	a := NewKVStore("node-a")
	b := NewKVStore("node-b")

	ua := a.Set("k", "from-a")
	ub := b.Set("k", "from-b")

	a.Apply(ub)
	b.Apply(ua)

	va, _ := a.Get("k")
	vb, _ := b.Get("k")
	if va != vb {
		t.Fatalf("stores diverged: a=%q b=%q", va, vb)
	}
	if va != "from-b" {
		t.Fatalf("converged on %q, want %q (node-id tiebreak)", va, "from-b")
	}
}

// Compare-and-swap is the primitive that makes a lock possible: the write must
// land only when the key still holds the version the caller saw.
func TestCompareAndSwap(t *testing.T) {
	kv := NewKVStore("self")
	first := kv.Set("lock", "alice")

	if _, ok := kv.Write("lock", "bob", WriteOptions{Expect: &pb.Version{Counter: 99, Node: "x"}}); ok {
		t.Fatal("CAS succeeded against the wrong version")
	}
	if v, _ := kv.Get("lock"); v != "alice" {
		t.Fatalf("value = %q after failed CAS, want alice", v)
	}
	if _, ok := kv.Write("lock", "bob", WriteOptions{Expect: &first.Version}); !ok {
		t.Fatal("CAS against the current version failed")
	}
	if v, _ := kv.Get("lock"); v != "bob" {
		t.Fatalf("value = %q after successful CAS, want bob", v)
	}
}

// A zero expected version means "the key must not exist" — claim semantics.
func TestCompareAndSwapRequiresAbsence(t *testing.T) {
	kv := NewKVStore("self")
	zero := pb.Version{}

	if _, ok := kv.Write("lock", "alice", WriteOptions{Expect: &zero}); !ok {
		t.Fatal("claiming a free key failed")
	}
	if _, ok := kv.Write("lock", "bob", WriteOptions{Expect: &zero}); ok {
		t.Fatal("claimed a key that was already held")
	}
	if v, _ := kv.Get("lock"); v != "alice" {
		t.Fatalf("holder = %q, want alice", v)
	}
}

func TestCompareAndDelete(t *testing.T) {
	kv := NewKVStore("self")
	u := kv.Set("k", "v")

	if _, ok := kv.Remove("k", WriteOptions{Expect: &pb.Version{Counter: 42}}); ok {
		t.Fatal("guarded delete succeeded against the wrong version")
	}
	if _, ok := kv.Remove("k", WriteOptions{Expect: &u.Version}); !ok {
		t.Fatal("guarded delete failed against the current version")
	}
	if _, ok := kv.Get("k"); ok {
		t.Fatal("key still present after guarded delete")
	}
}

// clockAt drives a store's notion of now, so TTL and GC are testable without
// sleeping.
func clockAt(kv *KVStore, t *time.Time) {
	kv.SetClock(func() time.Time { return *t })
}

func TestTTLHidesExpiredKey(t *testing.T) {
	now := time.Unix(1000, 0)
	kv := NewKVStore("self")
	clockAt(kv, &now)

	kv.Write("session", "token", WriteOptions{ExpiresAt: now.Add(10 * time.Second).Unix()})
	if _, ok := kv.Get("session"); !ok {
		t.Fatal("key missing before expiry")
	}

	now = now.Add(11 * time.Second)
	if _, ok := kv.Get("session"); ok {
		t.Fatal("expired key still readable")
	}
	if kv.Len() != 0 {
		t.Fatalf("Len = %d, want 0 with only an expired key", kv.Len())
	}
}

// Expiry must not bump the Lamport clock: a sweep that outranked a concurrent
// legitimate write would silently destroy it.
func TestSweepExpiredKeepsVersion(t *testing.T) {
	now := time.Unix(1000, 0)
	kv := NewKVStore("self")
	clockAt(kv, &now)
	u, _ := kv.Write("session", "token", WriteOptions{ExpiresAt: now.Add(time.Second).Unix()})

	now = now.Add(2 * time.Second)
	if got := kv.SweepExpired(); got != 1 {
		t.Fatalf("swept %d, want 1", got)
	}
	updates := kv.UpdatesFor([]string{"session"})
	if len(updates) != 1 {
		t.Fatalf("updates = %d, want 1", len(updates))
	}
	if updates[0].Action != pb.KVDelete {
		t.Fatalf("action = %s, want delete", updates[0].Action)
	}
	if !updates[0].Version.Equal(u.Version) {
		t.Fatalf("version = %+v, want the pre-expiry %+v", updates[0].Version, u.Version)
	}

	// A later write must still win over the swept tombstone.
	kv.Set("session", "fresh")
	if v, ok := kv.Get("session"); !ok || v != "fresh" {
		t.Fatalf("post-expiry write lost: (%q, %v)", v, ok)
	}
}

func TestGCTombstones(t *testing.T) {
	now := time.Unix(1000, 0)
	kv := NewKVStore("self")
	clockAt(kv, &now)
	kv.Set("a", "1")
	kv.Delete("a")
	kv.Set("b", "2")

	if got := kv.GCTombstones(time.Hour); got != 0 {
		t.Fatalf("GC reclaimed %d fresh tombstones, want 0", got)
	}
	now = now.Add(2 * time.Hour)
	if got := kv.GCTombstones(time.Hour); got != 1 {
		t.Fatalf("GC reclaimed %d, want 1", got)
	}
	if kv.Tombstones() != 0 {
		t.Fatalf("tombstones = %d after GC, want 0", kv.Tombstones())
	}
	if _, ok := kv.Get("b"); !ok {
		t.Fatal("GC removed a live key")
	}
}

// The digest walks the keyspace in batches; the reported range is what lets the
// receiver spot keys the sender never mentioned.
func TestDigestPaginates(t *testing.T) {
	kv := NewKVStore("self")
	for _, k := range []string{"a", "b", "c", "d"} {
		kv.Set(k, k)
	}

	first := kv.Digest("", 2)
	if len(first.Entries) != 2 || first.Entries[0].Key != "a" || first.Entries[1].Key != "b" {
		t.Fatalf("first batch = %+v", first.Entries)
	}
	if first.Hi == "" {
		t.Fatal("truncated digest must report an upper bound")
	}

	second := kv.Digest("b", 2)
	if len(second.Entries) != 2 || second.Entries[0].Key != "c" {
		t.Fatalf("second batch = %+v", second.Entries)
	}
	if second.Hi != "" {
		t.Fatalf("final batch Hi = %q, want empty (end of keyspace)", second.Hi)
	}
}

// Reconcile is the heart of anti-entropy: it must ask for what it lacks and
// offer what the peer lacks, in one comparison.
func TestReconcilePushesAndPulls(t *testing.T) {
	local := NewKVStore("local")
	local.Apply(pb.KVUpdate{Action: pb.KVSet, Key: "shared", Value: "new", Version: pb.Version{Counter: 5, Node: "x"}})
	local.Apply(pb.KVUpdate{Action: pb.KVSet, Key: "only-local", Value: "v", Version: pb.Version{Counter: 1, Node: "x"}})
	local.Apply(pb.KVUpdate{Action: pb.KVSet, Key: "stale-local", Value: "old", Version: pb.Version{Counter: 1, Node: "x"}})

	// The peer advertises the whole keyspace.
	d := pb.Digest{
		From: "peer",
		Entries: []pb.KeyVersion{
			{Key: "shared", Version: pb.Version{Counter: 2, Node: "x"}},
			{Key: "stale-local", Version: pb.Version{Counter: 9, Node: "x"}},
			{Key: "only-peer", Version: pb.Version{Counter: 1, Node: "x"}},
		},
	}
	push, pull := local.Reconcile(d, 100, 100)

	pushed := map[string]bool{}
	for _, u := range push {
		pushed[u.Key] = true
	}
	if !pushed["shared"] || !pushed["only-local"] {
		t.Fatalf("push = %v, want shared + only-local", keysOf(push))
	}
	if pushed["stale-local"] {
		t.Fatal("pushed an entry the peer holds a newer version of")
	}
	want := map[string]bool{"stale-local": true, "only-peer": true}
	if len(pull) != 2 || !want[pull[0]] || !want[pull[1]] {
		t.Fatalf("pull = %v, want stale-local + only-peer", pull)
	}
}

// A key outside the digest's advertised range must not be pushed: the sender
// simply had not reached that part of the keyspace yet.
func TestReconcileRespectsDigestRange(t *testing.T) {
	local := NewKVStore("local")
	local.Set("a", "1")
	local.Set("m", "2")
	local.Set("z", "3")

	d := pb.Digest{From: "peer", Lo: "", Hi: "n", Entries: []pb.KeyVersion{
		{Key: "a", Version: pb.Version{Counter: 99, Node: "x"}},
	}}
	push, _ := local.Reconcile(d, 100, 100)
	for _, u := range push {
		if u.Key == "z" {
			t.Fatalf("pushed %q from outside the digest range: %v", u.Key, keysOf(push))
		}
	}
	if len(push) != 1 || push[0].Key != "m" {
		t.Fatalf("push = %v, want just m", keysOf(push))
	}
}

func TestHistoryRecordsWriters(t *testing.T) {
	kv := NewKVStore("self")
	kv.Set("k", "v1")
	kv.Apply(pb.KVUpdate{Action: pb.KVSet, Key: "k", Value: "v2", Version: pb.Version{Counter: 50, Node: "peer"}})

	h := kv.History("k")
	if len(h) != 2 {
		t.Fatalf("history = %d entries, want 2", len(h))
	}
	if !h[0].Local {
		t.Fatal("first entry should be a local write")
	}
	if h[1].Local {
		t.Fatal("second entry should be a remote write")
	}
	if h[1].Value != "v2" {
		t.Fatalf("history value = %q, want v2", h[1].Value)
	}
}

func TestHistoryRingIsBounded(t *testing.T) {
	kv := NewKVStore("self")
	for i := 0; i < historyPerKey*2; i++ {
		kv.Set("k", "v")
	}
	if got := len(kv.History("k")); got != historyPerKey {
		t.Fatalf("history = %d entries, want %d", got, historyPerKey)
	}
}

func keysOf(updates []pb.KVUpdate) []string {
	out := make([]string, 0, len(updates))
	for _, u := range updates {
		out = append(out, u.Key)
	}
	return out
}
